tools/base/libboard/board_vc/board_vc.c

```diff
@@ -0,0 +1,749 @@
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <glob.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <net/ethernet.h>
+#include <opae/properties.h>
+#include <opae/utils.h>
+#include <opae/fpga.h>
+#include <sys/ioctl.h>
+
+#include "safe_string/safe_string.h"
+#include "board_vc.h"
+
+// sysfs paths
+#define SYSFS_BMCFW_VER                     "spi-altera.*.auto/spi_master/spi*/spi*.*/bmcfw_flash_ctrl/bmcfw_version"
+#define SYSFS_MAX10_VER                     "spi-altera.*.auto/spi_master/spi*/spi*.*/max10_version"
+#define SYSFS_PCB_INFO                      "spi-altera.*.auto/spi_master/spi*/spi*.*/pcb_info"
+#define SYSFS_PKVL_POLL_MODE                "spi-altera.*.auto/spi_master/spi*/spi*.*/pkvl/polling_mode"
+#define SYSFS_PKVL_STATUS                   "spi-altera.*.auto/spi_master/spi*/spi*.*/pkvl/status"
+#define SYSFS_BS_ID                         "bitstream_id"
+#define SYSFS_PHY_GROUP_INFO                "misc/eth_group*.*"
+#define SYSFS_PHY_GROUP_INFO_DEV            "misc/eth_group*/dev"
+#define SYSFS_EEPROM                        "*i2c*/i2c*/*/eeprom"
+#define SYSFS_NVMEM                         "*i2c*/i2c*/*/nvmem"
+#define SYSFS_PKVL_A_VER                    "spi-altera.*.auto/spi_master/spi*/spi*.*/pkvl/pkvl_a_version"
+#define SYSFS_PKVL_B_VER                    "spi-altera.*.auto/spi_master/spi*/spi*.*/pkvl/pkvl_b_version"
+
+// driver ioctl id
+#define FPGA_PHY_GROUP_GET_INFO               0xB702
+
+// Read BMC firmware version
+fpga_result read_bmcfw_version(fpga_token token, char *bmcfw_ver, size_t len)
+{
+	fpga_result res                = FPGA_OK;
+	fpga_result resval             = FPGA_OK;
+	uint32_t size                  = 0;
+	char buf[VER_BUF_SIZE]         = { 0 };
+	fpga_object bmcfw_object;
+
+	if (bmcfw_ver == NULL) {
+		FPGA_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_BMCFW_VER, &bmcfw_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to get token object");
+		return res;
+	}
+
+	res = fpgaObjectGetSize(bmcfw_object, &size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object size ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	// Return error if input sysfs path length smaller then object size
+	if (len < size) {
```

Comment:
I'm confused as to the relationship between `len` and `size` in this function. Is `len` here really the length of the sysfs path length? I don't think so.

Reply:
Fixed